catroid/src/main/java/org/catrobat/catroid/stage/EmbroideryActor.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.stage;
+
+import android.graphics.PointF;
+
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.graphics.g2d.Batch;
+import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
+import com.badlogic.gdx.scenes.scene2d.Actor;
+
+import org.catrobat.catroid.embroidery.EmbroideryManager;
+
+import java.util.ArrayList;
+import java.util.ListIterator;
+
+public class EmbroideryActor extends Actor {
+
+	@Override
+	public void draw(Batch batch, float parentAlpha) {
+		ArrayList<PointF> stitchPoints = EmbroideryManager.getStitchPoints();
+		ListIterator<PointF> iterator = stitchPoints.listIterator();
+
+		if (stitchPoints.size() >= 2) {
+			batch.end();
+
+			ShapeRenderer renderer = StageActivity.stageListener.shapeRenderer;
+			renderer.setColor(Color.BLACK);
+			renderer.begin(ShapeRenderer.ShapeType.Filled);
+
+			while (iterator.hasNext()) {
+				PointF point = iterator.next();
+				if (!iterator.hasNext()) {
+					renderer.circle(point.x, point.y, EmbroideryManager.getStitchSize());
+					renderer.end();
+					break;
+				}
+				PointF nextPoint = stitchPoints.get(iterator.nextIndex());
+				renderer.circle(point.x, point.y, EmbroideryManager.getStitchSize());
+				renderer.rectLine(point.x, point.y, nextPoint.x, nextPoint.y, EmbroideryManager.getStitchSize());
+			}
+
+			batch.begin();
+		}
+	}
+
+	public void reset() {
+		EmbroideryManager.clearStitchPoints();
+	}
+
+	public void dispose() {
+		EmbroideryManager.clearStitchPoints();
```

Comment:
this could potentially blow up when draw() is running.
(dispose is called from uithread, while draw is called from inside libgdx's render thread.)